File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/DraftAutosaveVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The shared draft table (`spec/draft-vectors.json`): the composer's edits in, the writes this client owes
 * the core out. The Vue, Flutter and SwiftUI kits run the same file through the same driver.
 */
class DraftAutosaveVectorsTest {
    @Test fun usesTheTablesDelay() {
        val rules = (FlareJson.parse(vectorsFile().readText()) as Map<*, *>)["rules"] as Map<*, *>
        assertEquals(FLARE_DRAFT_SAVE_DELAY_MS, (rules["saveDelayMs"] as Number).toInt())
    }

    @Test fun everyScriptWritesTheSameThingHere() = runEveryScript(0L)

    /** The same scripts with the clock where a real one is — see [TypingSignalVectorsTest]. */
    @Test fun everyScriptWritesTheSameThingWithARealClock() = runEveryScript(1_767_000_000_000L)

    private fun runEveryScript(base: Long) {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        val cases = table["cases"] as List<*>
        assertTrue(cases.size >= 23, "the shared table lost cases: ${cases.size}")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val name = case["name"] as String
            val drafts = FlareDraftAutosave()
            val saves = mutableListOf<FlareDraftSave>()
            for (rawStep in case["steps"] as List<*>) {
                val step = rawStep as Map<*, *>
                val at = base + (step["at"] as Number).toLong()
                saves += drafts.tick(at)
                val cid = step["conversationId"] as? String ?: ""
                val text = step["text"] as? String ?: ""
                when (step["op"] as String) {
                    "seed" -> drafts.seed(cid, text)
                    "edit" -> saves += drafts.edit(cid, text, at)
                    "send" -> saves += drafts.send(cid, at)
                    "restore" -> saves += drafts.restore(cid, text, at)
                    "leave" -> saves += drafts.leave(at)
                }
            }
            saves += drafts.tick(base + (case["endAt"] as Number).toLong())
            val expected = (case["expect"] as List<*>).map {
                val e = it as Map<*, *>
                FlareDraftSave(e["conversationId"] as String, e["text"] as String, base + (e["at"] as Number).toLong())
            }
            assertEquals(expected, saves, "$name @ base=$base")
        }
    }
}

/** The table lives beside the four kits, so find the repository root from the module's directory. */
private fun vectorsFile(): File {
    var dir: File? = File(".").absoluteFile
    while (dir != null) {
        val candidate = File(dir, "spec/draft-vectors.json")
        if (candidate.isFile) return candidate
        dir = dir.parentFile
    }
    error("spec/draft-vectors.json not found above ${File(".").absolutePath}")
}

File: packages/flutter-im-ui/test/draft_autosave_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared draft table (`spec/draft-vectors.json`): the composer's edits in, the writes this client
/// owes the core out. The Vue, SwiftUI and Compose kits run the same file through the same driver.
void main() {
  final table = jsonDecode(File('../../spec/draft-vectors.json').readAsStringSync()) as Map<String, dynamic>;
  final cases = (table['cases'] as List).cast<Map<String, dynamic>>();

  test('uses the table\'s delay, so the four kits cannot drift apart quietly', () {
    expect(flareDraftSaveDelayMs, (table['rules'] as Map<String, dynamic>)['saveDelayMs']);
  });

  for (final vector in cases) {
    final name = vector['name'] as String;
    for (final base in const [0, 1767000000000]) {
      test('$name${base == 0 ? '' : ' (on a real clock)'}', () {
        final drafts = FlareDraftAutosave();
        final saves = <FlareDraftSave>[];
        for (final raw in (vector['steps'] as List).cast<Map<String, dynamic>>()) {
          final at = base + (raw['at'] as int);
          saves.addAll(drafts.tick(at));
          final cid = raw['conversationId'] as String? ?? '';
          final text = raw['text'] as String? ?? '';
          switch (raw['op'] as String) {
            case 'seed':
              drafts.seed(cid, text);
            case 'edit':
              saves.addAll(drafts.edit(cid, text, at));
            case 'send':
              saves.addAll(drafts.send(cid, at));
            case 'restore':
              saves.addAll(drafts.restore(cid, text, at));
            case 'leave':
              saves.addAll(drafts.leave(at));
          }
        }
        saves.addAll(drafts.tick(base + (vector['endAt'] as int)));
        final expected = (vector['expect'] as List)
            .cast<Map<String, dynamic>>()
            .map((e) => FlareDraftSave(e['conversationId'] as String, e['text'] as String, base + (e['at'] as int)))
            .toList();
        expect(saves, expected, reason: name);
      });
    }
  }
}
